<!doctype html>
<head>
    <title>IBD Exomes Browser</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel=stylesheet type=text/css href="/static/typeaheadjs.css">
    <link rel=stylesheet type=text/css href="/static/css/font-awesome.min.css">
    <script type="text/javascript" src="/static/jquery-1.11.1.min.js"></script>
    <script type="text/javascript" src="/static/bootstrap.min.js"></script>
    <script type="text/javascript" src="/static/typeahead.bundle.min.js"></script>
    <script type="text/javascript" src="/static/jquery.tablesorter.min.js"></script>
    <script type="text/javascript" src="/static/jquery.tablesorter.pager.js"></script>
    <link rel=stylesheet type=text/css href="/static/theme.default.css">
    <script type="text/javascript" src="/static/underscore-min.js"></script>
    <link rel=stylesheet type=text/css href="/static/bootstrap.min.css">
    <link rel=stylesheet type=text/css href="/static/style.css">
    <script type="text/javascript" src="/static/d3.v3.min.js"></script>
    <script type="text/javascript" src="https://gabelerner.github.io/canvg/rgbcolor.js"></script>
    <script type="text/javascript" src="https://gabelerner.github.io/canvg/StackBlur.js"></script>
    <script type="text/javascript" src="https://gabelerner.github.io/canvg/canvg.js"></script>
    <script type="text/javascript" src="/static/index.js"></script>
    <script type="text/javascript" src="/static/exac.js"></script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-68395741-1', 'auto');
  ga('send', 'pageview');
</script>
    <script type="text/javascript">
        number_of_samples = 60706;
        release_number = 0.3;
        number_of_samples_full = 91796;
        $(document).ready(function() {
            $('.number_samples').html(Number(number_of_samples).toLocaleString('en'));
            $('.number_samples_full').html(Number(number_of_samples_full).toLocaleString('en'));
            $('.release_number').html(Number(release_number).toLocaleString('en'));
        });
        $(function() {
            var bestPictures = new Bloodhound({
              datumTokenizer: Bloodhound.tokenizers.obj.whitespace('value'),
              queryTokenizer: Bloodhound.tokenizers.whitespace,
              remote: '/autocomplete/%QUERY'
            });

            bestPictures.initialize();

            $('.awesomebar').typeahead(
                {
                    autoselect: true,
                },
                {
                    name: 'best-pictures',
                    displayKey: 'value',
                    source: bestPictures.ttAdapter(),
                }
            );
            $('.awesomebar').bind('typeahead:selected', function(obj, datum) {
                window.location.href = '/awesome?query=' + datum.value;
            });
        });
    </script>
</head>
<body>
<nav class="navbar navbar-default" role="navigation" style="background: #663399;">
    <div class="container-fluid">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle pull-right" data-toggle="collapse" data-target="#navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand" href="/" style="color: white; font-weight: bold; float: left; font-size: 15px;">IBD Exomes Browser Alpha</a>
            <div class="col-xs-5" id="navbar_form_container">
                <form action="/awesome" class="navbar-form" role="search">
                    <div class="form-group" id="navbar-awesomebar">
                        <input type="submit" style="display: none;"/>
                        <input id="navbar-searchbox-input" name="query" class="form-control awesomebar" type="text" placeholder="Gene, transcript, variant, or region"/>
                        <input type="submit" style="position: absolute; left: -9999px"/>
                    </div>
                </form>
            </div>
        </div>
        <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse" id="navbar-collapse">
            <ul class="nav navbar-nav navbar-right">
                <li><a href="/about" style="color: white; font-weight: bold;">About</a></li>
                <li><a href="/downloads" style="color: white; font-weight: bold;">Downloads</a></li>
                <li><a href="/terms" style="color: white; font-weight: bold;">Terms</a></li>
                <li><a href="/contact" style="color: white; font-weight: bold;">Contact</a></li>
                <li><a href="/faq" style="color: white; font-weight: bold;">FAQ</a></li>
            </ul>
        </div><!-- /.navbar-collapse -->
    </div><!-- /.container-fluid -->
</nav>




    <!-- Render context vars in JS here -->
    <script type="text/javascript">
        window.gene = {{ gene|tojson|safe }};
        window.transcript = {{ transcript|tojson|safe }};
        window.table_variants = {{ variants_in_gene|tojson|safe }};
        window.variants_in_transcript = {{ variants_in_transcript|tojson|safe }};
        window.transcripts_in_gene = {{ transcripts_in_gene|tojson|safe }};
        window.coverage_stats = {{ coverage_stats|tojson|safe }};
        window.page_name = window.gene.gene_id;
    </script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-68395741-1', 'auto');
  ga('send', 'pageview');

</script>
    <style>
div.transbox
{
  background-color: #ffffff;
  border: 2px solid gray;

}
    </style>
    <script>
        $(document).ready(function() {
            $("#variants_loading").hide();
            $("#variants_table_container").show();
        });
    </script>
    <div class="container-fluid">
        <div class="col-md-10 col-xs-offset-1 col-md-offset-1">
            <h1 style="color:black">Gene: NOD2</h1>
            <hr/>
        </div>
        <div class="row">
            <div class="col-md-6 col-xs-10 col-xs-offset-1 col-md-offset-0">
                <dl class="dl-horizontal">
                    
                        <dt>Trait</dt>
                        <dd>CD</dd>
                    
                    
                        <dt>NOD2</dt>
                        <dd>nucleotide-binding oligomerization domain containing 2</dd>
                    
                    <dt>Number of variants</dt>
                    <dd><span id="number_of_variants"></span> (Including filtered: 297)</dd>


                    <script>
                        filtered_variants = _.filter(window.table_variants, function(d) {
                            return d.filter == 'PASS';
                        });
                        $('#number_of_variants').replaceWith(filtered_variants.length);
                        lof_filtered_variants = _.filter(window.table_variants, function(d) {
                            return d.filter == 'PASS' && d.category == 'lof_variant';
                        });
                        $('#number_of_lof_variants').replaceWith(lof_filtered_variants.length);
                    </script>


                    <dt class="hidden-xs">UCSC Browser</dt>
                    <dd class="hidden-xs">
                        <a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&position=chr16%3A50727514-50766988" target="_blank">
                            16:50727514-50766988
                            <i class="fa fa-external-link"></i>
                        </a>
                    </dd>
                    <dt class="hidden-xs">ExAC</dt>
                    <dd class="hidden-xs">
                        <a href="http://exac.broadinstitute.org/gene/ENSG00000167207" target="_blank">
                            NOD2
                            <i class="fa fa-external-link"></i>
                        </a>
                    </dd>
                    
                        <dt>OMIM</dt>
                        <dd>
                            <a href="http://omim.org/entry/605956" target="_blank">
                                NOD2
                                <i class="fa fa-external-link"></i>
                            </a>
                        </dd>
                    
                    <dt>Other</dt>
                    <dd>
                        <div class="dropdown">
                            <button class="btn btn-default dropdown-toggle" type="button" id="external_ref_dropdown" data-toggle="dropdown">
                                External References
                                <span class="caret"></span>
                            </button>
                            <ul class="dropdown-menu" role="menu" aria-labelledby="external_ref_dropdown">
                                <li role="presentation" class="visible-xs-block">
                                    <a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&position=chr16%3A50727514-50766988" target="_blank">
                                        UCSC Browser
                                        <i class="fa fa-external-link"></i>
                                    </a>
                                </li>
                                <li role="presentation" class="visible-xs-block">
                                    <a href="http://www.genecards.org/cgi-bin/carddisp.pl?gene=NOD2" target="_blank">
                                        GeneCards
                                        <i class="fa fa-external-link"></i>
                                    </a>
                                </li>
                                <li role="presentation">
                                    <a role="menuitem" tabindex="-1" href="http://grch37.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000167207" target="_blank">
                                        Ensembl <i class="fa fa-external-link"></i>
                                    </a>
                                </li>
                                <li role="presentation">
                                    <a role="menuitem" tabindex="-1" href="http://en.wikipedia.org/NOD2" target="_blank">
                                        Wikipedia <i class="fa fa-external-link"></i>
                                    </a>
                                </li>
                                <li role="presentation">
                                    <a role="menuitem" tabindex="-1" href="http://www.ncbi.nlm.nih.gov/pubmed?term=NOD2" target="_blank">
                                        PubMed Search <i class="fa fa-external-link"></i>
                                    </a>
                                </li>
                                <li role="presentation">
                                    <a role="menuitem" tabindex="-1" href="http://www.wikigenes.org/?search=NOD2" target="_blank">
                                        Wikigenes <i class="fa fa-external-link"></i>
                                    </a>
                                </li>
                                <li role="presentation">
                                    <a role="menuitem" tabindex="-1" href="http://www.gtexportal.org/home/gene/NOD2" target="_blank">
                                        GTEx (Expression) <i class="fa fa-external-link"></i>
                                    </a>
                                </li>
                            </ul>
                        </div>
			</dd>
                </dl>
            </div>
	    <div class="col-md-5 hidden-xs">
                
		      <div class="panel-group" id="accordion">
			<div class="panel panel-default">
			  <div class="panel-heading">
			    <h4 class="panel-title">
			      
			    <a data-toggle="collapse" data-parent="#accordion" href="#collapse1">IBD LOF (CMH p = 8.597e-37)</a>
			    
			  </h4>
			</div>
			  <div id="collapse1" class="panel-collapse collapse">
			    <div class="panel-body">
			  <table cellspacing="5" cellpadding="5" style="width:100%;border:1px solid #ddd; 
    border-collapse: collapse;">
		  <tr>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">Population</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">Nvar</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">CNTA</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">CNTU</th>
		   </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">AJ</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">8</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">279</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">186</td>
		    </tr>
		  
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">FIN</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">7</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">50</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">341</td>
		    </tr>
		  
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">NFE</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">13</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">402</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">114</td>
		    </tr>
		  
		    </table> 
		  
			    
           <b>P-value: </b>8.597e-37 
		   <b>Odds ratio: </b>2.261
		   
		  
			    </div>
			  </div>
			</div>
			<div class="panel panel-default">
			  <div class="panel-heading">
			    <h4 class="panel-title">
			      
			    <a data-toggle="collapse" data-parent="#accordion" href="#collapse2">CD LOF (CMH p = 8.83e-63)</a>
			    
			  </h4>
			</div>
			  <div id="collapse2" class="panel-collapse collapse">
			    <div class="panel-body">
			  <table cellspacing="5" cellpadding="5" style="width:100%;border:1px solid #ddd; 
    border-collapse: collapse;">
		  <tr>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">Population</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">Nvar</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">CNTA</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">CNTU</th>
		   </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">AJ</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">7</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">247</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">186</td>
		    </tr>
		  
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">FIN</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">7</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">28</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">341</td>
		    </tr>
		  
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">NFE</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">11</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">344</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">114</td>
		    </tr>
		  
		    </table> 
		  
			    
			<b>P-value: </b>8.83e-63 
		   <b>Odds ratio: </b>3.073
		   
		  
			    </div>
			  </div>
			</div>

			<div class="panel panel-default">
			  <div class="panel-heading">
			    <h4 class="panel-title">
			      
			    <a data-toggle="collapse" data-parent="#accordion" href="#collapse3">UC LOF (CMH p = 0.6434)</a>
			    
			  </h4>
			</div>
			  <div id="collapse3" class="panel-collapse collapse">
			    <div class="panel-body">
			  <table cellspacing="5" cellpadding="5" style="width:100%;border:1px solid #ddd; 
    border-collapse: collapse;">
		  <tr>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">Population</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">Nvar</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">CNTA</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">CNTU</th>
		   </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">AJ</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">4</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">32</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">186</td>
		    </tr>
		  
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">FIN</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">6</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">17</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">341</td>
		    </tr>
		  
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">NFE</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">7</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">48</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">114</td>
		    </tr>
		  
		    </table> 
			  
			    
           <b>P-value: </b>0.6434 
		   <b>Odds ratio: </b>0.9419
		   
		  
			    </div>
			  </div>
			</div>
		
	    
			<div class="panel panel-default">
			  <div class="panel-heading">
			    <h4 class="panel-title">
			    <a data-toggle="collapse" data-parent="#accordion" href="#collapse4"><b>GWAS hit: fine-mapping info</b></a>
			  </h4>
			</div>
			  <div id="collapse4" class="panel-collapse collapse">
			    <div class="panel-body">
	      <p><strong>IN CD GWAS REGION 
	  
		AND DIRECTLY IMPLICATED
	   
		</strong>
		<br><a href="http://finemapping.broadinstitute.org/?HD=NOD2"><strong>Genes in region:</strong> CYLD,NOD2,SNX20</a>
		<br>
<!--		<table cellspacing="5" cellpadding="5" style="width:100%;border:1px solid #ddd; 
    border-collapse: collapse;">
		  <tr>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">SNP</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">AF</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">OR</th>
		    <th style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">P</th>
		   </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">rs5743293 (+0)</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">0.036</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">3.26</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">0</td>
		    </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">rs2066844 (+0)</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">0.063</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">2.00</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">1.32E-140</td>
		    </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">rs2066845 (+0)</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">0.022</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">2.25</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">2.29E-94</td>
		    </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">rs72796367 (+0)</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">0.023</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">1.47</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">1.95E-21</td>
		    </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">rs181708647 (+2)</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">0.004</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">2.23</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">7.08E-16</td>
		    </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">rs9673419 (+6)</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">0.249</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">1.11</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">9.55E-12</td>
		    </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">rs104895444 (+1)</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">0.003</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">2.02</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">2.04E-09</td>
		    </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">rs5743271 (+0)</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">0.007</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">1.51</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">1.45E-08</td>
		    </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">rs138425259 (+1)</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">0.009</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">1.45</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">2.29E-08</td>
		    </tr>
		  
                  <tr>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">rs104895467 (+2)</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">0.002</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">1.99</td>
 		    <td style="border-right:1px solid #ddd; border-bottom:1px solid #ddd;">1.20E-06</td>
		    </tr>
		  		   
		  <caption align="bottom" style="font-size:10px; text-align:left">IIBDGC results: Top variant in the fine-mapping credible set (+n others in the credible set). AF: allele frequency of variant; OR: odds ratio; P: p-value. </caption>
		    </table> -->
		</p>
			    </div>
			  </div>
			</div>
		      
		      </div>
            </div>
          </div>         
        </div>
        <div class="row">
            <div class="col-md-10 col-md-offset-1">
                <h3>Gene summary</h3>
                <div class="row">
                    
                        <div id="gene_plot">
    <span class="hidden-xs">
        <label for="display_coverage_metric_group">
            Display:
        </label>
        <span class="btn-group" data-toggle="buttons" id="display_coverage_metric_group">
            <button class="btn btn-primary active display_coverage_metric_buttons"
                    id="display_coverage_overview_button" data-tooltip="Shows all exons">
                <input type="radio">Overview</input>
            </button>
            <button class="btn btn-primary display_coverage_metric_buttons"
                    id="display_coverage_detail_button" data-tooltip="Zoomed in view (can scroll horizontally if needed)">
                <input type="radio">Detail</input>
            </button>
        </span>
        <label>
            <input type="checkbox" id="include_utrs_checkbox" value=""> Include UTRs in plot
        </label>
    </span>
     <span style="float:right;" class="hidden-xs">
    <span style="float: right;">
        <label for="coverage_metric_group">
            Association statistic:
        </label>
        <span class="btn-group" data-toggle="buttons" id="coverage_metric_group">
            <button class="btn btn-primary active coverage_metric_buttons"
                    id="covered_mean_covmet_button" data-tooltip="-log10(P-value) &#xa;threshold. Capped at 20">
                    <input type="radio">-log10(P-value)</input>
            </button>
            <button class="btn btn-primary coverage_metric_buttons" id="covered_covmet_button"
                data-tooltip="Odds ratio calculated using GLM.">
                <input type="radio">Odds Ratio</input> 
            </button>
        </span>
    </span>
    <div class="coverage_subcat_selectors" id="over_x_select_container" style="display: none; float:right;">
        <label for="over_x_select" style = "display : none">
            Odds Ratio
        </label>
        <select id="over_x_select" style = "display : none">
            <option selected>odds_ratio</option>
        </select>
    </div>
    <div  class="coverage_subcat_selectors" id="average_select_container" style="display: none; float:right;">
        <label for="average_select" style = "display : none">
            -log10(P-value):
        </label>
        <select id="average_select" style = "display : none">
            <option selected>-log10pvalue</option>
        </select>
    </div>

    <div  class="coverage_subcat_selectors" id="average_select_container" style="display: none; float:right;">
        <label for="average_select" style = "display : none">
            Log Odds ratio:
        </label>
       <select id="average_select" style = "display : none">
            <option selected>log_odds_ratio</option>
            <option>odds_ratio</option>
        </select>
    </div>  

</span>

    <br/>
    <div id="gene_plot_container_container">
        <div id="loading_coverage">
            Loading coverage...
            <br/><br/>
        </div>
        <span id="gene_plot_axis_container" class="hidden-xs" style="width:50px; display: inline-block;"></span>
        <span id="gene_plot_container" style="overflow-x:scroll; display: inline-block;"></span>
        <a class="btn btn-success" id="coverage_plot_download" target="_blank" download="NOD2_coverage">Save association plot</a>
        <a class="btn btn-success" id="exon_plot_download" target="_blank" download="NOD2_exon">Save exon image</a>
    </div>
</div>
<div id="not_covered" style="display: none;">
    No coverage for this transcript.
</div>
<br/>
<script>
function gene_chart(data, new_data, variant_data, _transcript) {
    var coords = 'pos_coding_noutr';
    var coding_coordinate_params = get_coding_coordinate_params(_transcript, true);
    var chart_width = gene_chart_width;
    var metric = '-log10pvalue';
    var metricor = 'odds_ratio';
    var exon_x_scale = d3.scale.linear()
        .domain([0, coding_coordinate_params.size])
        .range([0, chart_width]);
    var maxn = d3.max(data, function(d) { return d[metric] }) + 2;
    var maxnor = d3.max(data, function(d) { return d[metricor] }) + .3;
    var minor =  d3.min(data, function(d) { return d[metricor] }) - .3;
    var minor = minor < 0 ? 0 : minor; 
    var max_cov = (metric == '-log10pvalue') ? maxn : maxnor;
    var min_cov = (metric == '-log10pvalue') ? 0 : minor;
    var y = d3.scale.linear()
        .domain([min_cov, max_cov])
        .range([gene_chart_height, 0]);

    var yAxis = d3.svg.axis()
        .scale(y)
        .orient("left");

    var svg = d3.select('#gene_plot_container').append("svg")
        .attr("width", chart_width + gene_chart_margin.left + gene_chart_margin.right)
        .attr("height", gene_chart_height + gene_chart_margin.top + gene_chart_margin.bottom)
        .attr('id', 'inner_svg')
        .attr('class', 'hidden-xs')
        .append("g")
        .attr('id', 'inner_graph')
        .attr("transform", "translate(" + gene_chart_margin.left + "," + gene_chart_margin.top + ")");

    var area = d3.svg.area()
        .x( function(d) {
            return exon_x_scale(d[coords]);
        }).y0( function(d) {
            return gene_chart_height;
        }).y1( function(d) {
            return (metric in d) ? y(d[metric]) : gene_chart_height;
        });

    svg.append("g")
        .attr("class", "y axis")
        .call(yAxis);

    svg.selectAll("circle")
        .data(data)
        .enter()
        .append("circle")
        .attr("class", "dot")
 <!-- stuff i added -->
        .attr("data-toggle", "tooltip")
        .attr('filter_status', function(d) {
            return d.flag;
        })
        .attr('category', function(d) {
            return d.category;
        })
        .attr("class", function(d) {
            return "track_variant " + d.category;
        })
 <!-- stuff i added -->
        .attr("cx", function(d) {
            if (d['pos_coding_noutr'] == undefined) {
                return -100;
            } else {
                return exon_x_scale(d[coords]);
            }
        })
        .attr("cy", function(d){
               return (metric in d) ? y(d[metric]) : gene_chart_height;
        })
        .attr('r', 4 )
        .style("fill", variant_colors);

    d3.select('#gene_plot_container').append("br"); //make sure track_svg is below graph and not to the right of it

    // plot exons
    var svg_outer = d3.select('#gene_plot_container').append("svg")
        .attr("width", chart_width + gene_chart_margin_lower.left + gene_chart_margin_lower.right)
        .attr("height", lower_gene_chart_height)
        .attr('id', 'track_svg')
        .append("g")
        .attr('id', 'track')
        .attr("transform", "translate(" + gene_chart_margin_lower.left + "," + 0 + ")");
    var exon_color = "lightsteelblue";
    svg_outer.append("line")
        .attr("y1", lower_gene_chart_height/2)
        .attr("y2", lower_gene_chart_height/2)
        .attr("x1", 0)
        .attr("x2", exon_x_scale(coding_coordinate_params.size))
        .attr('id', 'boundary_line')
        .attr("stroke-width", 5)
        .attr("stroke", exon_color);

    console.log(_transcript);
    // plot exon rects
    svg_outer.selectAll("bar")
        .data(_transcript.exons)
        .enter()
        .append("rect")
        .attr('class', 'track_bar')
        .style("fill", exon_color)
        .attr("x", function(d, i) { return exon_x_scale(get_coding_coordinate(_transcript, d.start, true)); })
        .attr("y", function(d, i) {
            if (d.feature_type == 'CDS') {
                return 0;
            } else {
                return lower_gene_chart_height/4;
            }
        })


        .attr("width", function(d, i) {
            if (get_coding_coordinate(_transcript, d.start, true) == -100) {
                return exon_x_scale(175);
            }
            return exon_x_scale(d.stop-d.start+1);
        })
        .attr("height", function(d, i) {
            if (d.feature_type == 'CDS') {
                return lower_gene_chart_height;
            } else {
                return lower_gene_chart_height/2;
            }
        });


    var a_s = _transcript.strand == "-"? -1 : 1; //arrow direction
    var a_x = -5;  //arrow position on x-axis
    var a_y = lower_gene_chart_height/2.0; //arrow position on y-axis
    var a_w = 2; //arrow width
    var points = [[a_x+a_s*6, a_y], [a_x+a_s*1, a_y+a_w*3], [a_x+a_s*1, a_y+a_w], [a_x-a_s*9, a_y+a_w],
        [a_x-a_s*9, a_y-a_w], [a_x+a_s*1, a_y-a_w], [a_x+a_s*1, a_y-a_w*3]];
    svg_outer.append("polygon")
            .attr("points", points.join(" "))
            .attr("fill", "steelblue")
            .attr("stroke", "black");

    var bounds = get_af_bounds(variant_data);
    var min_af = bounds[0];
    var max_af = bounds[1];
    var min_af = .01;
    var max_af = 1;
    var variant_size_scale = d3.scale.log()
        .domain([min_af, max_af])
        //Circle/Ellipse
        .range([2, lower_gene_chart_height/3]);
        //Rectangle
//        .range([lower_gene_chart_height, 2]);

    // show variant category on hover
    var tip = d3.tip().attr('class', 'd3-tip').html(function(d) {
        if (d.category) {
            var csq = d.major_consequence.replace('_variant', '')
                    .replace('_', ' ')
                    .replace('utr', 'UTR')
                    .replace('3 prime', "3'")
                    .replace('5 prime', "5'")
                    .replace('nc ', "non-coding ");
            var output = csq + '<br/>' + d.chrom + ':' + d.pos + ' ' + d.ref + '&#8594;' + d.alt;
            if (d.major_consequence == 'missense_variant' || d.major_consequence == 'synonymous_variant') {
                output += '<br/>' + d.HGVSp;
            }
            //output += '<br/>P-value: ' + d.ibd_meta_p.toPrecision(3);
            //output += '<br/>Odds Ratio: ' + d.ibd_meta_or.toPrecision(3);
            return output;
        } else {
            return 'None';
        }
    });
    svg.call(tip);

    // plot variants
    svg_outer.selectAll("bar")
        .data(variant_data)
        .enter()
        .append("a")
        .attr('class', 'track_variant_link')
        .attr("xlink:href", function(d, i) { return "/variant/" + d.chrom + "-" + d.pos + "-" + d.ref + "-" + d.alt; })
        .attr("data-toggle", "tooltip")
        .attr('filter_status', function(d) {
            return d.filter;
        })
        .attr('category', function(d) {
            return d.category;
        })
        .attr('variant_id', function(d) {
            return d.variant_id;
        })
        .on('mouseover', function(d) {
            $('#variant_' + d.variant_id).find('td').addClass('table_hover');
            tip.show(d);
        })
        .on('mouseout', function(d) {
            $('#variant_' + d.variant_id).find('td').removeClass('table_hover');
            tip.hide(d);
        })
        //Circle
//        .append("circle")
        //Ellipse
        .append("ellipse")
        .attr("class", function(d) {
            return "track_variant " + d.category;
        })
        .style("opacity", 0.9)
        .attr("cx",  function(d) {
            if (d['pos_coding_noutr'] == undefined) {
                return -100;
            } else {
                return exon_x_scale(d[coords]);
            }
        })
        .attr("cy", lower_gene_chart_height/2)
        //Circle
//        .attr("r", function(d, i) { return variant_size_scale(d.ibd_meta_p); })
        //Ellipse
        .attr("rx", 2)
        .attr("ry", function(d, i) {
            if (d.ibd_meta_p == 0) {
                                return (2*(variant_size_scale(max_af) - variant_size_scale(.0000000000000000001)));
            } else {
               if (d.ibd_meta_p){
                return (2*(variant_size_scale(max_af) - variant_size_scale(d.ibd_meta_p)));
                }
              else { 
                 return 0;
                  }
            }
        })
        // Workaround for exporting d3 to SVG (other delcaration in style.css).
        .attr('stroke', variant_colors)
        .attr('fill', variant_colors);
        //Rectangle
//        .append("rect")
//        .attr("class", "track_variant")
//        .style("fill", "darkred")
//        .style("opacity", 0.5)
//        .attr("x", function(d, i) {
//            var tx_coord = d.transcript_coordinates[transcript];
//            if (tx_coord == 0) {
//                return -1000;
//            } else {
//                var variant_exon_number = d.vep_annotations[0]['EXON'].split('/')[0] - 1;
//                return exon_x_scale(tx_coord + variant_exon_number*padding);
//            }
//        })
//        .attr("y", function(d, i) { return lower_gene_chart_height/2 - variant_size_scale(d.ibd_meta_p)/2; } )
//        .attr("width", 2)
//        .attr("height", function(d, i) { return variant_size_scale(d.ibd_meta_p); })
//        .attr("rx", 6)
//        .attr("ry", 6);

}

function variant_colors(d) {
    if (d.category == 'lof_variant') {
        return '#cd2932';
    } else if (d.category == 'missense_variant') {
        return '#FF9966';
    } else if (d.category == 'synonymous_variant') {
        return '#228B22';
    }
    else{ 
         return "#d3d3d3";
    }
}

function change_coverage_chart(data, new_data, variant_data, _transcript, scale_type, metric, skip_utrs, container) {
    var coords = skip_utrs ? 'pos_coding_noutr' : 'pos_coding';
    var metricor = 'odds_ratio'; 
    var maxn = d3.max(data, function(d) { return d[metric] }) + 2;
    var maxnor = d3.max(data, function(d){ return d[metricor]}) + .3;
    var minor =  d3.min(data, function(d) { return d[metricor] }) - .3;
    var minor = minor < 0 ? 0 : minor; 
    var max_cov = (metric == '-log10pvalue') ? maxn : maxnor;
    var min_cov = (metric == '-log10pvalue') ? 0 : minor;
    var coding_coordinate_params = get_coding_coordinate_params(_transcript, skip_utrs);
    var chart_width;
    if (scale_type == 'overview') {
        chart_width = gene_chart_width;
    } else {
        chart_width = coding_coordinate_params.size*2;
    }

    var exon_x_scale = d3.scale.linear()
        .domain([0, coding_coordinate_params.size])
        .range([0, chart_width]);

    var svg = d3.select(container).select('#inner_svg')
        .attr("width", chart_width + gene_chart_margin.left + gene_chart_margin.right)
        .attr("height", gene_chart_height + gene_chart_margin.top + gene_chart_margin.bottom)
        .select('#inner_graph');

    var y = d3.scale.linear()
        .domain([min_cov, max_cov])
        .range([gene_chart_height, 0]);

    var area = d3.svg.area()
        .x( function(d) {
            return exon_x_scale(d[coords]);
        }).y0( function(d) {
            return gene_chart_height;
        }).y1( function(d) {
            return (metric in d) ? y(d[metric]) : gene_chart_height;
        });

    svg.selectAll("circle")
        .data(data)
        .transition()
        .duration(500)
        .attr("class", "dot")
<!-- stuff i added -->
        .attr("data-toggle", "tooltip")
        .attr('filter_status', function(d) {
            return d.flag;
        })
        .attr('category', function(d) {
            return d.category;
        })
        .attr("class", function(d) {
            return "track_variant " + d.category;
        })
 <!-- stuff i added -->

        .attr("cx", function(d) {
            if (d['pos_coding_noutr'] == undefined) {
                return -100;
            } else {
                return exon_x_scale(d[coords]);
            }
        })
        .attr("cy", function(d){
               return (metric in d) ? y(d[metric]) : gene_chart_height;
        })
        .attr('r', 4 )
        .style("fill", variant_colors);
;

    // plot exons
    var svg_outer = d3.select(container).select('#track_svg')
        .attr("width", chart_width + gene_chart_margin_lower.left + gene_chart_margin_lower.right)
        .attr("height", lower_gene_chart_height).select('#track');

    svg_outer.select('#boundary_line')
        .attr("x2", exon_x_scale(coding_coordinate_params.size));

    // plot exon rounded rects
    svg_outer.selectAll("rect")
        .data(_transcript.exons)
        .transition()
        .duration(500)
        .attr("x", function(d, i) { return exon_x_scale(get_coding_coordinate(_transcript, d.start, skip_utrs)); })
        .attr("width", function(d, i) {
            if (get_coding_coordinate(_transcript, d.start, skip_utrs) == -100) {
                return exon_x_scale(175);
            }
            return exon_x_scale(d.stop-d.start+1);
        })
        .attr("height", function(d, i) {
            if (d.feature_type == 'CDS') {
                return lower_gene_chart_height;
            } else {
                return lower_gene_chart_height/2;
            }
        });

    // plot variants
    svg_outer.selectAll("a")
        .data(variant_data)
        .transition()
        .duration(500)
        .selectAll('ellipse')
        .attr("cx", function(d) {
            if (d[coords] == undefined) {
                return -100;
            } else {
                return exon_x_scale(d[coords]);
            }
        });

    var yAxis = d3.svg.axis()
        .scale(y)
        .orient("left");

    svg.select(".y.axis")
        .transition()
        .duration(200)
        .call(yAxis);
}

function create_new_data(data, coords) {
    var data_object = {};
    $.each(data, function(i, d) {
        data_object[d[coords]] = d;
    });
    var new_data = [];
    for (var i = d3.min(data, function(d) { return d[coords] }); i < d3.max(data, function(d) { return d[coords] }); i++) {
        var x = {'has_coverage': false};
        x[coords] = i;
        //Check the previous base to see if this is the beginning of an exon

        if (i in data_object && !(i-1 in data_object)) {
            new_data.push(x);
        }
        //Check the previous base to see if this is the end of an exon

        if (!(i in data_object) && i-1 in data_object) {
            x[coords] = i-1;
            new_data.push(x);
        }
        if (i in data_object) {
            new_data.push(data_object[i]);
        } else {
            new_data.push(x);
        }
    }
    return new_data;
}

function coverage_sum(key) {
    var total = 0;
    $.map(window.coverage_stats, function(entry) {
        total += entry[key];
    });
    return (total/window.coverage_stats.length).toPrecision(4);
}

function refresh_links() {
    $("#coverage_plot_download").attr('href', set_plot_image('gene_plot_container', 0));
    $("#exon_plot_download").attr('href', set_plot_image('gene_plot_container', 1));
}

$(document).ready(function() {
    if ($(window).width() < 768) {
        $('#gene_plot_container').css('width', $(window).width() + "px");
    } else {
        $('#gene_plot_container').css('width', $(window).width()*10/12 + "px");
    }
    precalc_coding_coordinates(window.transcript, window.coverage_stats, 'pos');
    precalc_coding_coordinates(window.transcript, window.variants_in_transcript, 'pos');

    // only show variants that have a coding coordinate
    window.variants_in_transcript = _.filter(window.variants_in_transcript, function(variant) {
        return variant.pos_coding != undefined;
    });

    // only show coding rects that have a coding coordinate
    window.coverage_stats = _.filter(window.coverage_stats, function(d) {
        return d.pos_coding != undefined;
    });
    $('#avg_coverage').html(coverage_sum('-log10pvalue'));
    $('#avg_coverage_x').html(coverage_sum('30')*100 + '%');

    var new_data = create_new_data(window.coverage_stats, 'pos_coding');
    var new_data_skip_utr = create_new_data(window.coverage_stats, 'pos_coding_noutr');
    if (window.coverage_stats != null) {
        gene_chart(window.coverage_stats, new_data_skip_utr, window.variants_in_transcript, window.transcript);
        if (window.variants_in_transcript.length) {
            update_variants();
        }
        $('#loading_coverage').hide();
    } else {
        $('#gene_plot').hide();
        $('#not_covered').show();
    }
    // Change coverage plot
    $('.coverage_metric_buttons').change(function () {
        var v = $(this).attr('id').replace('_covmet_button', '');
        $('.coverage_subcat_selectors').hide();
        if (v == 'covered') {
           $('#over_x_select_container').show();
            v = $('#over_x_select').val().replace('X', ''); 
        } else {
            $('#average_select_container').show();
            v = $("#average_select").val();
        }
        var detail = $('.display_coverage_metric_buttons.active').attr('id').replace('display_coverage_', '').replace('_button', '');
        var include_utrs = $('#include_utrs_checkbox').is(':checked');
        var plot_data = include_utrs ? new_data : new_data_skip_utr;
        change_coverage_chart(window.coverage_stats, plot_data, window.variants_in_transcript, window.transcript, detail, v, !include_utrs, '#gene_plot_container');
        refresh_links();
    });
    $('#over_x_select').change(function () {
        var detail = $('.display_coverage_metric_buttons.active').attr('id').replace('display_coverage_', '').replace('_button', '');
        var include_utrs = $('#include_utrs_checkbox').is(':checked');
        var plot_data = include_utrs ? new_data : new_data_skip_utr;
        $('#avg_coverage_type_x').html($(this).val());
        $('#avg_coverage_x').html(coverage_sum($(this).val().replace('X', ''))*100 + '%');
        change_coverage_chart(window.coverage_stats, plot_data, window.variants_in_transcript, window.transcript, detail, $(this).val().replace('X', ''), !include_utrs, '#gene_plot_container');
        refresh_links();
    });
    $('#average_select').change(function () {
        var detail = $('.display_coverage_metric_buttons.active').attr('id').replace('display_coverage_', '').replace('_button', '');
        var include_utrs = $('#include_utrs_checkbox').is(':checked');
        var plot_data = include_utrs ? new_data : new_data_skip_utr;
        $('#avg_coverage_type').html($(this).val());
        $('#avg_coverage').html(coverage_sum($(this).val()));
        change_coverage_chart(window.coverage_stats, plot_data, window.variants_in_transcript, window.transcript, detail, $(this).val(), !include_utrs, '#gene_plot_container');
        refresh_links();
    });
    $('#include_utrs_checkbox').change(function () {
        setTimeout(function() {
            var detail = $('.display_coverage_metric_buttons.active').attr('id').replace('display_coverage_', '').replace('_button', '');
            var v = $('.coverage_metric_buttons.active').attr('id').replace('_covmet_button', '');
            v = (v == 'covered') ? $('#over_x_select').val().replace('X', '') : $("#average_select").val(); 
            var include_utrs = $('#include_utrs_checkbox').is(':checked');
            var plot_data = include_utrs ? new_data : new_data_skip_utr;
            change_coverage_chart(window.coverage_stats, plot_data, window.variants_in_transcript, window.transcript, detail, v, !include_utrs, '#gene_plot_container');
            refresh_links();
        }, 10);
    });

// Change exon diagram
    $('#inverted_checkbox').change(function () {
        setTimeout(function () {
            var v = $('#inverted_checkbox').is(':checked');
            change_track_chart_variant_size(window.variants_in_transcript, v, '#gene_plot_container');
            refresh_links();
        }, 10);
    });

    $('.consequence_display_buttons, #filtered_checkbox').change(function () {
        setTimeout(function() {
            update_variants();
            refresh_links();
        }, 10);
    });

    $('.display_coverage_metric_buttons').change(function () {
        var detail = $(this).attr('id').replace('display_coverage_', '').replace('_button', '');
        var v = $('.coverage_metric_buttons.active').attr('id').replace('_covmet_button', '');
        if (v == 'covered') {
            $('#over_x_select_container').show();
            v = $('#over_x_select').val().replace('X', ''); 

        } else {
            $('#average_select_container').show();
            v = $("#average_select").val();
        }
        var include_utrs = $('#include_utrs_checkbox').is(':checked');
        var plot_data = include_utrs ? new_data : new_data_skip_utr;
        change_coverage_chart(window.coverage_stats, plot_data, window.variants_in_transcript, window.transcript, detail, v, !include_utrs, '#gene_plot_container');
        refresh_links();
    });
    refresh_links();
});
</script>
                    
                </div>
                <div class="row">
                    
                        
                        
<span id="variant_selectors_container">
    <label for="variant_consequence_types">
        <span class="btn-group" data-toggle="buttons" id="variant_consequence_types">
            <button class="btn btn-primary consequence_display_buttons active" id="consequence_other_variant_button">
                <input type="radio" >All
            </button>
            <button class="btn btn-primary consequence_display_buttons" id="consequence_missense_variant_button">
                <input type="radio" >Missense + LoF
            </button>
            <button class="btn btn-primary consequence_display_buttons" id="consequence_lof_variant_button">
                <input type="radio">LoF
            </button>
        </span>
    </label>
    <label class='hidden-xs'>
        <input type="checkbox" id="filtered_checkbox" value=""> Include filtered (non-PASS) variants
    </label>
<span id="pop_selector_types" class="btn-group" data-toggle="buttons" style="float:right;">
  <button type="button" class="btn btn-primary pop_selector active" name="AJ">AJ</button>
    <button type="button" class="btn btn-primary pop_selector" name="FIN">FIN</butto\
n>
    <button type="button" class="btn btn-primary pop_selector" name="NFE">NFE</button>
</span>

</span>
<span style="float:right">
</span>
<script type="text/javascript">
    window.is_gene = typeof window.transcripts_in_gene != 'undefined';
</script>
<script type="text/template" id="variant-table-template">
<span>
    <a class="btn btn-success hidden-xs" id="export_to_csv">Export table to CSV</a>
</span>

    <span style="float:left">
        <small>&dagger; denotes a consequence that is for a non-canonical transcript</small>
    </span>

<table id="variant_table" class="tablesorter">
    <thead>
        <tr>
            <th class="tooltip-table-header omit_csv" data-tooltip="chr:pos reference/alternate &#xa;(dbSNP135 rsID)">
                Variant</th>
            <th class="hidden-xs tooltip-table-header" data-tooltip="Population enrichment: ratio comparing non-IBD allele frequency in Ashkenazi population to NFE (non-Finnish European) in ExAC &#xa;">Enrichment</th>
            <th class='hidden-xs'>Position</th>
            <td class="hidden">RSID</td>
            <th class="hidden">Reference</th>
            <th class="hidden">Alternate</th>
            <th class="hidden-xs tooltip-table-header" data-tooltip="HGVS annotation (protein change, where defined, or transcript change for splice variants, otherwise empty)">Consequence</th>
            <th class="hidden">Protein Consequence</th>
            <th class="hidden">Transcript Consequence</th>
            <th class="hidden-xs tooltip-table-header" data-tooltip="VQSR sensitivity filter & hard filters">Filter</th>
            <th class="tooltip-table-header" data-tooltip="Variant Effect Predictor (VEP) &#xa; annotation using Gencode 75.&#xa; Worst across all transcripts of this gene.">Annotation</th>
            <th class="tooltip-table-header" data-tooltip="ExAC NFE alternate allele frequency &#xa;">
                ExAC NFE AF</th>
            <th class="tooltip-table-header" data-tooltip="IBD alternate allele frequency &#xa;">
                IBD AF</th>
            <th class="hidden-xs tooltip-table-header" data-tooltip="non-IBD alternate allele frequency &#xa;">
                non-IBD AF</th>
            
                <th class="hidden-xs tooltip-table-header" data-tooltip="Posterior Odds ratio">Odds Ratio</th>
            
            
<th class="tooltip-table-header allele_frequency_column" data-tooltip="Meta-analysis IBD association P-value &#xa; The graphic displays &#xa; p-value on a  &#xa; discrete scale: &#xa; <0.01, <0.001, < 1e-4, <1e-5">                
P-value</th>
            <th class="hidden">Allele Count African</th>
            <th class="hidden">Allele Number African</th>
            <th class="hidden">Homozygote Count African</th>
            <th class="hidden">Allele Count East Asian</th>
            <th class="hidden">Allele Number East Asian</th>
            <th class="hidden">Homozygote Count East Asian</th>
            <th class="hidden">Allele Count European (Non-Finnish)</th>
            <th class="hidden">Allele Number European (Non-Finnish)</th>
            <th class="hidden">Homozygote Count European (Non-Finnish)</th>
            <th class="hidden">Allele Count Finnish</th>
            <th class="hidden">Allele Number Finnish</th>
            <th class="hidden">Homozygote Count Finnish</th>
            <th class="hidden">Allele Count Latino</th>
            <th class="hidden">Allele Number Latino</th>
            <th class="hidden">Homozygote Count Latino</th>
            <th class="hidden">Allele Count Other</th>
            <th class="hidden">Allele Number Other</th>
            <th class="hidden">Homozygote Count Other</th>
            <th class="hidden">Allele Count South Asian</th>
            <th class="hidden">Allele Number South Asian</th>
            <th class="hidden">Homozygote Count South Asian</th>
        </tr>
    </thead>
    <tbody>
    <% _.each(table_variants, function(variant) { %>
        <tr class="table_variant" id="variant_<%= variant.variant_id %>"
            category="<%= variant.category %>" filter_status="<%= variant.filter %>" frequency="<%= variant.allele_freq %>">
            <td class="omit_csv">
                <a href="/variant/<%= variant.chrom %>-<%= variant.pos %>-<%= variant.ref %>-<%= variant.alt %>" target="_blank">
                <%= variant.chrom %>:<%= variant.pos %>
                <% if (variant.ref.length > 20) { %>
                    <%= variant.ref.slice(0,20) %>...
                <% } else { %>
                    <%= variant.ref %>
                <% } %> /
                <% if (variant.alt.length > 20) { %>
                    <%= variant.alt.slice(0,20) %>...
                <% } else { %>
                    <%= variant.alt %>
                <% } %>
                <% if (variant.rsid != '.') { %>
                    (<%= variant.rsid %>)
                <% } %>
                </a>
            </td>
<td>
</td>
            <td class='hidden-xs'> <%= variant.pos %></td>
            <td class="hidden"><%= variant.rsid %></td>
            <th class="hidden"> <%= variant.ref %></th>
            <th class="hidden"> <%= variant.alt %></th>
            <td class='hidden-xs'>
                <% if (variant.HGVSp != '') { %>
                    <%= variant.HGVSp %><% if (is_gene & variant.CANONICAL != 'YES') { %>&dagger;<% } %>
                <% } else { %>
                    <% if (variant.major_consequence == 'splice_donor_variant' | variant.major_consequence == 'splice_acceptor_variant' | variant.major_consequence == 'splice_region_variant') { %>
                        <% if (variant.HGVSc.length > 20) { %>
                            <%= variant.HGVSc.slice(0,20) %>...<% } else { %>
                            <%= variant.HGVSc %><% } %><% if (is_gene & variant.CANONICAL != 'YES') { %>&dagger;<% } %>
                    <% } %>
                <% } %>
            </td>
            <td class='hidden'> <%= variant.HGVSp %></td>
            <td class='hidden'> <%= variant.HGVSc %></td>
            <td class='hidden-xs'> <%= variant.filter %> </td>
            <td class="<%= variant.category %>" data-consequence="<%= variant.major_consequence %>"><b>
                <% if (variant.major_consequence) { %>
                    <%= variant.major_consequence.replace('_variant', '').replace(/_/g, ' ').replace('utr', 'UTR').replace('3 prime', "3'").replace('5 prime', "5'").replace('nc ', "non-coding ") %>
                <% } %>
            </b> </td>
            <td> 
                <% if (variant.exac_nfe != -9) { %>
                <% } else { %>

                <% } %>    
            </td>
            <td> 
	            <% if(window.selected_population  == "AJ" & variant.pop_acs["European (Ashkenazi-Jewish)"] != -9) { %>                                                                  
		          <%= variant.pop_acs["European (Ashkenazi-Jewish)"].toPrecision(3) %>                                                                 
              <% } else{ %>
              <% if(window.selected_population  == "AJ" & variant.ibd_calta_aj != -9){ %>
            <%= (variant.ibd_calta_aj/4378).toPrecision(3) %>
            <% } %>   
             <% if(window.selected_population  == "FIN" & variant.pop_acs["European (Finnish)"] != -9) { %>                                                                  
	           <%= variant.pop_acs["European (Finnish)"].toPrecision(3)  %>                                                          
              <% } else { %> 
             <% if(window.selected_population  == "FIN" & variant.ibd_calta_fin != -9){ %>
            <%= (variant.ibd_calta_fin/1610).toPrecision(3) %>
            <% } %>   
             <% if(window.selected_population  == "NFE" & variant.pop_acs["Non-Finnish European"] != -9) { %>                                                                  
	           <%= variant.pop_acs["Non-Finnish European"].toPrecision(3)  %>                                                          
              <% } else { %>     
              <%  if(window.selected_population == "NFE" & variant.ibd_calta_nfe != -9){ %>                        
                 <%= (variant.ibd_calta_nfe/6944).toPrecision(3) %>
                <% }  %> 
              <% }  %>
              <% } } %> 
            </td>
            <td class='hidden-xs'>
	            <% if(window.selected_population  == "AJ" & variant.pop_ans["European (Ashkenazi-Jewish)"] != -9) { %>                                                                  
		          <%= variant.pop_ans["European (Ashkenazi-Jewish)"].toPrecision(3) %>                                                                 
              <% } else{ %>
               <% if(window.selected_population  == "AJ" & variant.ibd_caltu_aj != -9){ %>
            <%= (variant.ibd_caltu_aj/5538).toPrecision(3) %>
            <% } %>   
             <% if(window.selected_population  == "FIN" & variant.pop_ans["European (Finnish)"] != -9) { %>                                                                  
	           <%= variant.pop_ans["European (Finnish)"].toPrecision(3)  %>                                                          
              <% } else { %>   
            <% if(window.selected_population  == "FIN" & variant.ibd_caltu_fin != -9){ %>
            <%= (variant.ibd_caltu_fin/23652).toPrecision(3) %>
            <% } %>                        
             <% if(window.selected_population  == "NFE" & variant.pop_ans["Non-Finnish European"] != -9) { %>                                                                  
	           <%= variant.pop_ans["Non-Finnish European"].toPrecision(3)  %>                                                          
              <% } else { %>  
              <%  if(window.selected_population == "NFE" & variant.ibd_caltu_nfe != -9){ %>                            
              <%= (variant.ibd_caltu_nfe/5044).toPrecision(3) %>
              <% }  %> 
              <% }  %>
              <% } } %> 
        </td>
            
                <td class="hidden-xs"> 
                <% if (variant.ibd_meta_or != -9) { %>
		 <% if (variant.ibd_meta_or == 0) { %>
                    <% '< 1e-16' %>
                <% } else { %> 
                <%= variant.ibd_meta_or.toPrecision(3) %>
                <%  } %>
		<% } else { %>
                    <% if (variant.ibd_meta_or != -9) { %>
                    <%= variant.ibd_meta_or.toPrecision(3) %>
                <% } else { %>
		<%= variant.ibd_cmhmeta_or.toPrecision(3) %> 
		<%  } %>
                <% } %>
                 </td>
            
            
            <td>
                <% if (variant.ibd_meta_p != -9) { %>
                    <% if (variant.ibd_meta_p < .00000000000001) { %>
                    <%= '< 1e-16' %>
                <% } else { %>
                <%= variant.ibd_meta_p.toPrecision(3) %>
                <% } %>
                <% } else { %>
                    <% if (variant.ibd_meta_p != -9) { %>
                    <%= variant.ibd_meta_p.toPrecision(3) %>
                <% } else { %>
                  <%= variant.ibd_cmhmeta_p.toPrecision(3) %> 
		<%  } %>
                <% } %>
                <div id="variant_af_box_<%= variant.variant_id %>" style="float: right;"></div>
            </td>
            <td class="hidden"> <%= variant.pop_acs["African"] %></td>
            <td class="hidden"> <%= variant.pop_ans["African"] %></td>
            <td class="hidden"> <%= variant.pop_homs["African"] %></td>
            <td class="hidden"> <%= variant.pop_acs["East Asian"] %></td>
            <td class="hidden"> <%= variant.pop_ans["East Asian"] %></td>
            <td class="hidden"> <%= variant.pop_homs["East Asian"] %></td>
            <td class="hidden"> <%= variant.pop_acs["European (Non-Finnish)"] %></td>
            <td class="hidden"> <%= variant.pop_ans["European (Non-Finnish)"] %></td>
            <td class="hidden"> <%= variant.pop_homs["European (Non-Finnish)"] %></td>
            <td class="hidden"> <%= variant.pop_acs["European (Finnish)"] %></td>
            <td class="hidden"> <%= variant.pop_ans["European (Finnish)"] %></td>
            <td class="hidden"> <%= variant.pop_homs["European (Finnish)"] %></td>
            <td class="hidden"> <%= variant.pop_acs["Latino"] %></td>
            <td class="hidden"> <%= variant.pop_ans["Latino"] %></td>
            <td class="hidden"> <%= variant.pop_homs["Latino"] %></td>
            <td class="hidden"> <%= variant.pop_acs["Other"] %></td>
            <td class="hidden"> <%= variant.pop_ans["Other"] %></td>
            <td class="hidden"> <%= variant.pop_homs["Other"] %></td>
            <td class="hidden"> <%= variant.pop_acs["South Asian"] %></td>
            <td class="hidden"> <%= variant.pop_ans["South Asian"] %></td>
            <td class="hidden"> <%= variant.pop_homs["South Asian"] %></td>
        </tr>
    <% }); %>
    </tbody>
</table>
</script>
<script>
  window.selected_population  = "AJ";
    var populate_variant_table = function() {

        // This list is simply copied from utils.py.
        // Obviously that's not ideal, but I wanted to keep this diff isolated.
        // I'm also not sure how you want to serve additional context to the client -
        // yet another template variable also seems suboptimal.
        // xBrowse has a variable DICTIONARY in global scope where we store stuff like this -
        // might want to try that!
        var csq_order = ["transcript_ablation",
            "splice_donor_variant",
            "splice_acceptor_variant",
            "stop_gained",
            "frameshift_variant",
            "stop_lost",
            "initiator_codon_variant",
            "transcript_amplification",
            "inframe_insertion",
            "inframe_deletion",
            "missense_variant",
            "splice_region_variant",
            "incomplete_terminal_codon_variant",
            "stop_retained_variant",
            "synonymous_variant",
            "coding_sequence_variant",
            "mature_miRNA_variant",
            "5_prime_UTR_variant",
            "3_prime_UTR_variant",
            "non_coding_transcript_exon_variant",
            "non_coding_exon_variant",
            "intron_variant",
            "NMD_transcript_variant",
            "non_coding_transcript_variant",
            "nc_transcript_variant",
            "upstream_gene_variant",
            "downstream_gene_variant",
            "TFBS_ablation",
            "TFBS_amplification",
            "TF_binding_site_variant",
            "regulatory_region_ablation",
            "regulatory_region_amplification",
            "regulatory_region_variant",
            "feature_elongation",
            "feature_truncation",
            "intergenic_variant",
        ""];

        // Make a map of consequence -> index
        // That'll be the actual sort key, so lower i => more severe consequence
        var csq_order_index = {};
        _.each(csq_order, function(c, i) {
            csq_order_index[c] = i;
        });

        // Here we add a new sort parser - its global identifier is 'consequence'
        $("#variants_loading").hide();
        $("#variants_table_container").show();
    window.variants_template = _.template($('#variant-table-template').html());
    // $('#variants_table_container').html(variants_template({table_variants: table_variants}));
        //define click event handler that re-renders the table using the selected population
          $("#variant_table").tablesorter({
        //    sortList: [[1,0], [2,0], [10,1]],
       // sortList: [[1,0], [1,0], [1,1]],
            // Here's where we tell tablesorter to use our new consequence sorter on the 9th column
            // 2-minute minor on Konrad for having hidden columns
            headers: {
                10: {
                    sorter: 'consequence'
                }
            }
        });


            $('#pager').hide();


        function get_af_category(d) {
            if (d.ibd_meta_p == 0) {
                return [7, '<1e-6'];
            } else if ((!d.ibd_meta_p) | (d.ibd_meta_p == -9)) {
			              if (d.ibd_cmhmeta_p == 0){
			               return [7, '<1e-6'];
						             } else {
                  if (d.ibd_cmhmeta_p <= 1/1000000){ 
                     return [7, '<1e-6'];
                  } else if (d.ibd_cmhmeta_p <= 1/100000) {
                     return [6, '<1e-5'];
                  } else if (d.ibd_cmhmeta_p <= 1/10000) {
                     return [5, '1e-4-1e-5'];
                  } else if (d.ibd_cmhmeta_p <= 1/1000) {
                     return [4, '1e-3-1e-4'];
                  } else if (d.ibd_cmhmeta_p <= 1/100) {
                     return [3, '0.01-0.001'];
                  } else if (d.ibd_cmhmeta_p <= 1/20) {
                     return [2, '0.05-0.01'];
                  } else {
                     return [1, '>0.05'];
                  } 
						       }
            } else if (d.ibd_meta_p <= 1/1000000){ 
                return [7, '<1e-6'];
            } else if (d.ibd_meta_p <= 1/100000) {
                return [6, '<1e-5'];
            } else if (d.ibd_meta_p <= 1/10000) {
                return [5, '1e-4-1e-5'];
            } else if (d.ibd_meta_p <= 1/1000) {
                return [4, '1e-3-1e-4'];
            } else if (d.ibd_meta_p <= 1/100) {
                return [3, '0.01-0.001'];
            } else if (d.ibd_meta_p <= 1/20) {
                return [2, '0.05-0.01'];
            } else {
                return [1, '>0.05'];
            } 
        }

        var data = window.table_variants;

        var width = 50;
        var height = 15;

        var x_scale = d3.scale.linear()
            .domain([0, 7])
            .range([0, width]);

        function get_posterior(d) {
				       if (d.posterior_null == -9){
                return [0, 0, 0, 0, 0];
				       }
				       else{
				       return [d.posterior_null, d.posterior_sim, d.posterior_het, d.posterior_cd, d.posterior_uc];
				       }
        }


        var data = window.table_variants;

        var width = 50;
        var height = 15;

        var x_scale = d3.scale.linear()
            .domain([0,7])
            .range([0, width]);

        var svg; 
        $.each(data, function(i, d) {
            d3.select('#variant_af_box_' + d.variant_id).attr("data-tooltip", "Shows ibd meta-analysis p-value \n on a discrete " +
                    "scale: \n <1e-6, <1e-5, \n <1e-4, <0.001, 0.01-0.001, \n 0.05-0.01, >0.05. \n This particular variant is \n " +
                    get_af_category(d)[1] + ".");
            svg = d3.select('#variant_af_box_' + d.variant_id)
                .append("svg")
                .attr("width", width)
                .attr("height", height)
                .append("g");

            for (var j=0; j<8; j++) {
                svg.append('rect')
                        .style('stroke', 'steelblue')
                        .style('fill', 'white')
                        .attr('x', x_scale(j))
                        .attr('y', 0)
                        .attr('height', height)
                        .attr('width', x_scale(1) - x_scale(0))
            }

            svg.append('rect')
                .style('fill', 'steelblue')
                .attr('x', 0)
                .attr('y', 0)
                .attr('width', function() {
                    return x_scale(get_af_category(d)[0]);
                })
                .attr('height', height);

        });

 
			       update_variants();
        $("#export_to_csv").on('click', function (event) {
            var output_name = window.page_name === undefined ? 'export' : window.page_name;
            var timestamp = date_format(new Date());
            exportTableToCSV.apply(this, [$('#variant_table'), 'exac_' + output_name + '_' + timestamp + '.csv']);
        });
    };
			       $(document).ready(populate_variant_table);
			        $('.pop_selector').on('click', function (event) {
              $('.pop_selector').addClass('active').not(event).removeClass('active');
               window.selected_population = event.target.name;
			       populate_variant_table();
    });
</script>
<div id="variants_loading">Loading variants...</div>
<div id="variants_table_container" style="display: none;"></div>
<div id="variants_table_empty" style="display: none;">No variants found.</div>

                    
                </div>
            </div>
        </div>
    </div>

</body>
